ExpressionUtils/Evaluating/CachedExpressionCompiler.cs

```diff
@@ -84,5 +86,27 @@ private static object getDefaultValue(Type t) {
 		internal bool IsCached(LambdaExpression lambda) {
 			return delegates.ContainsKey(lambda);
 		}
+
+		/// <summary>
+		/// Previously as a key for <see cref="delegates"/> we were using <see cref="LambdaExpression"/> objects and <see cref="ExpressionComparing.StructuralComparer"/> as a comparer.
+		/// But that required us to make calls to <see cref="Expression.Lambda(Expression, ParameterExpression[])"/> which contains global lock inside, and that started to become a problem.
+		/// So instead we now use <see cref="LambdaParts"/> as a key and pass body and parameters separately, to reduce number of calls to <see cref="Expression.Lambda(Expression, ParameterExpression[])"/>.
+		/// </summary>
+		private class LambdaExpressionComparer : IEqualityComparer<LambdaParts> {

```

Comment:
```suggestion
		private class LambdaPartsComparer : IEqualityComparer<LambdaParts> {
```
This seems more precise